datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -268,8 +271,7 @@ public void start() {
     // Initializing executor services
     _scheduledExecutor = Executors.newSingleThreadScheduledExecutor(
         new ThreadFactoryBuilder().setNameFormat("CoordinatorScheduledExecutor-%d").build());
-    // TODO Assess whether having a single threaded executor for token claim tasks is sufficient or it will be exhausted
-    _tokenClaimExecutor = Executors.newSingleThreadExecutor(
+    _tokenClaimExecutor = Executors.newFixedThreadPool(TOKEN_CLAIM_THREAD_POOL_SIZE,
```

Comment:
A single threaded token claim executor will get backed up in the following cases:
* The leader will try to schedule 2 tasks in this executor upon a stream stop: (1) a task for polling the Zookeeper to check whether stop has propagated and completed across the cluster (2) a task for claiming the token assigned to itself.
* Concurrent requests to stop different streams.

A single threaded executor won't work here. For now I increased the thread pool size to 8 threads. That will let the cluster process up to 8 concurrent requests without being backed up (note that token claim tasks are short lived unlike the poll tasks, so they won't be the bottleneck here). I'll monitor how this performs in EI and make adjustments accordingly. A config property can also be added in the future if we end up tinkering with this.

Reply:
great catch!